File: 0139-word-break/0139-word-break.cpp
/*

dp[i] =  Is it possible to  break [i....n-1] into dict

Ans= dp[0]

dp[i] = Is str[i....j] present in dict && Is it possible to break [j+1...n-1]


dp[i] = Is str[i....j] present in dict && dp[j+1]

for all j=i to n-1

*/


int dp[301];

class Solution {
public:
    unordered_map<string,int>mp;
    int n;
    bool wordBreak(string s, vector<string>& wordDict) {
        
        
        memset(dp,-1,sizeof(dp));
        
        for(string str:wordDict){
            mp[str]++;
        }
        
        n=s.size();
        
        return find(0,s);
          
    }
    
    
    
    bool find(int i,string &str){
        
      /// base case
        
        if(i>=n)
            return true;
        
        
        
        if(dp[i]!=-1)
            return dp[i];
        
       
       bool f1=false;
        
        
       bool f2=false;
        
        for(int j=i;j<n;j++){
        
            int l=j-i+1;
            
            string temp=str.substr(i,l);
            
            if(mp.count(temp)==1){
                
                f1=true;

                bool f3=true;
                
                f3=f3 & find(j+1,str);
                
                if(f3){
                f2=true;
                break;
                }
                
            }
            
        }
             
        if(f1)
            return dp[i]=f2;
        
    
        return dp[i]= false;
        
    }
};

File: 0139-word-break/0139-word-break.py


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
    
        n=len(s)
        
        m=len(wordDict)
        
        
        
        self.dp=[-1 for _ in range(n+1)]

        
        mp=defaultdict(int)
        
        for i in range(m):
            mp[wordDict[i]]+=1
            
            
            
        return self.find(0,n,s,mp)
    
    
    
    def find(self,i,n,s,mp):
        
        if i>=n:
            return True;
        
        if self.dp[i]!=-1:
            return self.dp[i]
        
        f1=False
        f2=False
        
        for j in range(i,n,1):
            if mp[s[i:j+1]]>0:
                f1=True
                f3=True
                f3=f3 & self.find(j+1,n,s,mp)
                
                if f3==True:
                    f2=True
                    break
        
        
        if f1==True:
            self.dp[i]=f2
            return self.dp[i]
        
        self.dp[i]=False
        return self.dp[i]
    
    